Metrics Type 2.0 (#5281)
DataDog/documentation: content/en/developers/metrics/metrics_type.md

---
title: Metrics Types
kind: documentation
aliases:
 - /developers/metrics/counts/
 - /developers/metrics/distributions/
 - /developers/metrics/gauges/
 - /developers/metrics/histograms/
 - /developers/metrics/rates/
 - /developers/metrics/sets/
further_reading:
- link: "developers/dogstatsd"
  tag: "Documentation"
  text: "Learn more about DogStatsD"
- link: "developers/libraries"
  tag: "Documentation"
  text: "Official and Community-contributed API and DogStatsD client libraries"
---

The "Datadog in-app type" affects how a given metric is interpreted in query results and graph visualizations across the application. This type is visible and can be changed on the [metric summary page][1]. Be aware that changing the metric type may render historical data nonsensical.

> In the Datadog web application there are five metric types (though one is deprecated):

Reviewer: ```suggestion
There are five metric types in the Datadog web application (though one is deprecated):
```

* COUNT
* COUNTER (deprecated)
* GAUGE
* RATE
* DISTRIBUTION

A metric's type is stored as metrics metadata and is used to determine how a metric is interpreted throughout the application by determining default time aggregation function and `as_rate()`/`as_count()` behavior. The `as_count()` and `as_rate()` modifiers behave differently for different web application metric types.

## Metric type definition.

> To better understand the different metrics types, what they represent and how to manipulate them within Datadog let's start with an example:

Reviewer: ```suggestion
To better understand the different metrics types, what they represent, and how to manipulate them within Datadog, let's start with an example:
```

> You have two web servers `web_1` and `web_2`, each one of them is dealing with HTTP requests over time.

Reviewer: ```suggestion
You have two web servers `web_1` and `web_2`. Each one of them deals with HTTP requests over time.
```

> They receive both:

Reviewer: ```suggestion
They both receive:
```

  * 1 request per second for 10 seconds
  * 2 requests per second for 10 seconds
  * 0 request for 10 seconds

> and so on and so fourth.

Reviewer: ```suggestion
```

{{< tabs >}}
{{% tab "Count" %}}

> **Metrics typed with `Count` are used to count things over a period of time.**

Reviewer: ```suggestion
**Metrics of type `Count` are used to count things over a period of time.**
```

For instance let's say the `number.of.requests` metrics is reported every 10 seconds to Datadog with the `count` type for `web_1`.

> The metric has then the following shape: 10, then 20 then 0

Reviewer: ```suggestion
The metric has then the following shape: 10, then 20, then 0.
```

Shape? Like, once it's graphed, you mean?

> Each data point of this metrics represent the amount of request received during the 10 second interval.

Reviewer: ```suggestion
Each data point represents the number of requests received during the 10 second interval.
```

{{% /tab %}}
{{% tab "Gauge" %}}

> **Gauges measure the value of a particular thing over time. It's a snapshot of a value associated to a timestamp.**

Reviewer: ```suggestion
**Gauges measure the value of a particular thing over time. It's a snapshot of a value associated with a timestamp.**
```

For instance let's say the `number.of.requests` metrics is reported every 10 seconds to Datadog with the `gauge` type for `web_1`.

> The metric has then the following shape: 10 , then 30 then 30.

Reviewer: ```suggestion
The metric has then the following shape: 10, then 30, then 30.
```

Shape? Like, once it's graphed, you mean?

{{% /tab %}}
{{% tab "Rate" %}}

> **Rates represent the derivative of a metric, it's the value variation of a metric on a defined time interval.**

Reviewer: ```suggestion
**Rates represent the derivative of a metric: the value variation of a metric on a defined time interval.**
```

For instance let's say the `number.of.requests` metrics is reported every 10 seconds to Datadog with the `gauge` type for `web_1`.

> The metric has then the following shape: 1 , then 2 then 0.

Reviewer: ```suggestion
The metric has then the following shape: 1, then 2, then 0.
```

Shape? Like, once it's graphed, you mean?

{{% /tab %}}

{{% tab "Distribution" %}}

<div class="alert alert-warning">
This feature is in beta. <a href="https://docs.datadoghq.com/help/">Contact Datadog support</a> to enable distribution metrics for your account.
</div>

Distributions are a metric type that aggregate values sent from multiple hosts during a flush interval to measure statistical distributions across your entire infrastructure. Distribution metrics are designed to instrument logical objects, like services, independently from the underlying hosts, and solve the problem created by Agent-level aggregation.

> Unlike the histogram metric type that aggregates on the Agent-side, distributions send all raw data collected during the flush interval and aggregations occur server-side. Because the underlying data structure has not been aggregated and represents raw data, distributions provide two major features:

Reviewer: ```suggestion
Unlike the histogram metric type that aggregates on the Agent-side, distributions send all raw data collected during the flush interval, and aggregations occur server-side. Because the underlying data structure has not been aggregated and represents raw data, distributions provide two major features:
```

* [COUNT](?tab=count#metric-type-definition)
* [RATE](?tab=rate#metric-type-definition)
* [GAUGE](?tab=gauge#metric-type-definition)
* [HISTOGRAM](?tab=histogram#metric-type-definition)
* [DISTRIBUTION](?tab=distribution#metric-type-definition)
* [SET](?tab=set#metric-type-definition)
* [TIMERS](?tab=timers#metric-type-definition)

Choosing a submission metric type depends on your use case and how your application is designed. Sometimes, incrementing a `COUNT` metric is easier than keeping track of a variable historical value that you would send with a `GAUGE` metric, and sometimes it's the other way around. If you want to combine multiple sources of submission—for example to calculate the average response time of different instances of a given service—then a `HISTOGRAM` or a `DISTRIBUTION` metric might be better suited. Refer to the [Metric Type Definition](#metric-type-definition) section below to learn more about the different metrics types available. Also, refer to the [Submission Types and Datadog In-App Types Section](#submission-types-and-datadog-in-app-types) to see how each metric type behaves between its submission and its storage within Datadog.

> ## Metric type definition

Author: Really good input.

* `0.66` for the second interval of 30 seconds
* `0` for the last interval of 30 seconds

Then this pattern starts again, since the `RATE` is the normalized per-second variation of the number of requests:

{{< img src="developers/metrics/metric_types/rate_metric.png" alt="Rate Metric" responsive="true">}}

Discover how to submit rate metrics:

* [With a custom Agent Check][1]
* [With the Datadog API][2]

**Note**: for a `RATE` metric, when a `0` value is submitted, `null` is stored within Datadog.

[1]: /developers/metrics/agent_metrics_submission/?tab=rate
[2]: /api/?lang=python#post-timeseries-points
{{% /tab %}}
{{% tab "GAUGE" %}}

**The `GAUGE` metric type represents the value of a particular thing over time.** It is a snapshot of a value associated with a timestamp.

For instance, say that the `number.of.requests.gauge` metric is reported every 30 seconds to Datadog with the `GAUGE` type for `server:web_1`.

Each data point represents the overall number of requests received at a point in time. The metric then has the following shape:

* `10` for the first 30 seconds
* `30` for the second interval of 30 seconds
* `30` for the last interval of 30 seconds

> Then this pattern starts again, but it increases over time since the `GAUGE` metric type keeps track of how many requests have already been received:

Reviewer 2: this is an interesting example, it really shows what happens when you use `gauge` for this type of data. Here, I think it show that it's non-sensical to track your example with a `gauge` (the increasing value in Datadog can't really be used for anything).

So maybe we should have another example where we would not could a number of requests, but a "state", for example the number of users currently browsing on our website (or the number of active connexions to our web servers, or something similar), where it makes sense to use gauges, and it generally doesn't make sense to use counts/rates.

(^ all this is only useful to explain the difference between gauges and rates/counts, but not really to explain what happens with histograms, distribution metrics, etc)

Author: Talked about this off-line, this is solved.
